Add documentation outlining the steps to transition from placementRul… (#491)
open-cluster-management-io/policy-collection: README.md

- [Policy Generator reference YAML](https://github.com/stolostron/policy-generator-plugin/blob/main/docs/policygenerator-reference.yaml)
- [Policy Generator examples](policygenerator)

## Using PlacementRule to distribute policies (Deprecated)

### Background

As PlacementRule has been **deprecated**, it's recommended to migrate towards using Placement for defining resource placement strategies within multi-cluster environments. This transition aims to adopt a more modern and versatile format offered by Placement.

### Objectives of Conversion

> The goal of converting from PlacementRule to Placement is to update your configuration from an older, deprecated format to a newer, more flexible one. This ensures alignment with current best practices and enables leveraging the enhanced capabilities provided by Placement.

Reviewer: I think all of this can be deleted.  The new sections you have are the main content and the following section provides a good start on the Overview.

[PlacementBinding](https://open-cluster-management.io/concepts/policy/#placementbinding) binds a Placement to a `Policy` or `PolicySet`.

[Placement](https://open-cluster-management.io/concepts/placement/) concept is used to dynamically select a set of `managedClusters` in one or multiple `ManagedClusterSet` so that higher level users can either replicate Kubernetes resources to the member clusters or run their advanced workload i.e. multi-cluster scheduling.

When the `Policy` is bound to a `Placement` using a `PlacementBinding`, the Policy status will report on each cluster that matches the bound Placement

### Using `Placement` with `Policies`

To use Placement for policies without having previously used PlacementRule, please refer to this [guide](https://open-cluster-management.io/concepts/placement/)

### Deprecated PlacementRule details

> **_NOTE:_** `PlacementRule` is now **deprecated**. We recommend switching to `Placement` instead.

> The policies are applied to all available managed clusters with the environment set to "dev." To apply the policies to a different set of clusters, update the PlacementRule.spec.clusterSelector.matchExpressions section in the policies.

Reviewer: You copied this from a previous section, but this information is wrong.  Please delete the previous content containing these sentences.
```suggestion
Policies previously defined the `PlacementRule` and `PlacementBinding` resources along with the `Policy` configuring the policy to be distributed to all managed clusters with the label `environment` set to "dev." To apply your policies to managed clusters, use the information provided in this section instead since the `PlacementRule` and `PlacementBinding` resources are no longer being included with the policies in this repository.
```

Author: Because if "Policies previously defined..."  Should this part be inserted after "Prerequisite"?

> As new clusters meeting the specified criteria are added, the policies will be applied to them automatically.

Reviewer: Probably should delete or move this sentence somewhere more appropriate.

### Prerequisite

- [ManagedClusterSet](https://open-cluster-management.io/concepts/managedclusterset/) and [ManagedClusterSetBinding](https://open-cluster-management.io/concepts/managedclusterset)
- `Policy`(The `namespace` in the `policy` should be bound to `ManagedClusterSetBinding`)

> - `PlacementBinding` and `PlacementRule`

Reviewer: It isn't clear why there's a Prerequisite section here.  I think you are wanting to show what needs to exist on the cluster prior to starting a procedure, but I think it's hard to understand without some more details.  I think it would make more sense to have a list of steps maybe.

Step 1: Make sure your clusters are organized as desired into ManagedCluster sets... see link 
Step 2: Make sure you have a ManagedClusterSetBinding for the namespace where you are authoring policies.  See link...
Step 3: Look for any PlacementBinding resources that point to a PlacementRule.  For each of those PlacementBindings, you need to continue with these steps.
Step 4: Change the PlacementBinding
Step 5: Change the PlacementRule

- [Policy Generator reference YAML](https://github.com/stolostron/policy-generator-plugin/blob/main/docs/policygenerator-reference.yaml)
- [Policy Generator examples](policygenerator)

## Distributing Policies to Managed Clusters

Distributing a `Policy` to a managed cluster requires four parts, all of which must be in the same namespace:

> - [Policy](https://open-cluster-management.io/concepts/policy/) is a grouping mechanism for Policy Templates and is the smallest deployable unit on the hub cluster. Embedded Policy Templates are distributed to applicable managed clusters and acted upon by the appropriate `policy controller`.

Reviewer 2: Lines such as this one are still too long--would you please run Prettier (or something similar) in VSCode to wrap them to match the rest of the file?

Author: I used both no change
![Screenshot 2024-07-18 at 4 47 06 PM](https://github.com/user-attachments/assets/20277514-b476-4a8d-a750-dd9f5ccb98c3)

Author: Where do you want to change?

Reviewer 2: Oh! You _might_ need to set the "Prettier: Print Width" to 100 (I think I usually have it set to 120, but it looks like this file was originally wrapped at 100)

![image](https://github.com/user-attachments/assets/2ee63168-81e5-4787-abd5-fb2c17109dfa)

Author: <img width="482" alt="Screenshot 2024-07-19 at 10 13 35 AM" src="https://github.com/user-attachments/assets/108705da-c3cb-4cde-9914-533022b0d08e">
Umm nothing changed

Reviewer 2: Oh, I think I know. Try setting "Prettier: Prose Wrap" to "always":
![image](https://github.com/user-attachments/assets/b7219e80-1d4d-4f70-87c6-1c6867ddad28)

Author: OH~~~ changed a lot

## Distributing Policies to Managed Clusters

Distributing a `Policy` to a managed cluster requires four parts, all of which must be in the same namespace:

- [Policy](https://open-cluster-management.io/concepts/policy/) is a grouping mechanism for Policy Templates and is the smallest deployable unit on the hub cluster. Embedded Policy Templates are distributed to applicable managed clusters and acted upon by the appropriate `policy controller`.

- [PlacementBinding](https://open-cluster-management.io/concepts/policy/#placementbinding) binds a Placement to a `Policy` or `PolicySet`.

- [Placement](https://open-cluster-management.io/concepts/placement/): Dynamically selects a set of `ManagedClusters` in one or multiple `ManagedClusterSet`s.

- [ManagedClusterSetBinding](https://open-cluster-management.io/concepts/managedclusterset/): Binds a `ManagedClusterSet` to a namespace, making this group of managed clusters available for selection by `Placement`.

When the `Policy` is bound to a `Placement` using a `PlacementBinding`, the `Policy` is distributed to the managed clusters and the `Policy` status will report on each cluster returned by the bound `Placement`.

> ### Using `Placement` with `Policies`
> 
> See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for additional details on selecting managed clusters using `Placement`.
> 
> ### Deprecated PlacementRule details
> 
> > **_NOTE:_** `PlacementRule` is now **deprecated**. We recommend switching to `Placement` instead.
> 
> To use `Placement` to distribute `Policies`, bind the `Policy` to the `Placement` using a `PlacementBinding`. All of the objects must be in the same namespace. View the following sample `Placement` and `PlacementBinding` bound to a `Policy` named `policy-example` in the namespace `example-ns`. The `Placement` selects managed clusters that have the label `environment: dev`:

Reviewer 2: These sections can be combined into a single heading.
I've also removed "now" (which is unclear since that could be any time it's being read) and "recommend" since it's a requirement and not optional given the deprecation.

```suggestion
### Using `Placement` with `Policies`

See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for additional details on selecting managed clusters using `Placement`.

- **NOTE:** `PlacementRule` is **deprecated**. See the [migration section](#transitioning-from-placementruledeprecated-to-placement) for detail on migrating to `Placement`.

To use `Placement` to distribute `Policies`, bind the `Policy` to the `Placement` using a `PlacementBinding`. All of the objects must be in the same namespace. View the following sample `Placement` and `PlacementBinding` bound to a `Policy` named `policy-example` in the namespace `example-ns`. The `Placement` selects managed clusters that have the label `environment: dev`:
```

### Deprecated PlacementRule details

> **_NOTE:_** `PlacementRule` is now **deprecated**. We recommend switching to `Placement` instead.

Previously, policies defined the `PlacementRule` and `PlacementBinding` resources, configuring the policy to be distributed to all managed clusters with the environment label set to "dev." To apply your policies to managed clusters, use the information provided in this section, as `PlacementRule` and `PlacementBinding` resources are **no longer included with the policies in this repository**.

As new clusters meeting the specified criteria are added, the policies will be applied to them automatically.

**Example of `PlacementRule` and `PlacementBinding`**:

```yaml
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
  name: binding-policy-example
placementRef:
  name: placement-policy-example
  kind: PlacementRule
  apiGroup: apps.open-cluster-management.io
subjects:
  - name: policy-example
    kind: Policy
    apiGroup: policy.open-cluster-management.io
---
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
  name: placement-policy-example
spec:

>   clusterSelector:
>     matchExpressions:
>       - { key: environment, operator: In, values: ["dev"] }
> ```
> 
> ### Transitioning from `PlacementRule`(deprecated) to `Placement`
> 
> #### Step 1:
> 
> Make sure your clusters are organized as desired into `ManagedClusterSet`. [See link](https://open-cluster-management.io/concepts/managedclusterset/)
> 
> #### Step 2:
> 
> Make sure you have a `ManagedClusterSetBinding` for the `namespace` where you are authoring policies. [See link](https://open-cluster-management.io/concepts/managedclusterset)
> 
> > **Caution**: The `namespace` under `ManagedClusterSetBinding` should match the namespace in your `Policy`
> 
> #### Step 3:
> 
> Identify any `PlacementBinding` resources that reference a `PlacementRule`. For each of those `PlacementBindings`, proceed with the following steps.
> 
> #### Step 4:
> 
> Create the `Placement` manifest as shown below. This `Placement` manifest replaces the `PlacementRule`. Ensure that the `namespace` in `Placement` matches the `namespace` used in `ManagedClusterSetBinding`. Copy `matchExpressions` from the `PlacementRule` and paste them into `spec.predicates.requiredClusterSelector.labelSelector`.
> 
> #### Step 5:
> 
> Update the `PlacementBinding` to reference the new `Placement`. Change the `placementRef.kind` to `Placement` and update the `placementRef.apiGroup` to `cluster.open-cluster-management.io` to reflect the `Placement`'s API version.
> 
> Expected Result:
> 
> ```yaml
> apiVersion: cluster.open-cluster-management.io/v1beta1
> kind: Placement
> metadata:
>   name: placement-policy-example
>   namespace: example-ns // Please ensure that this namespace matches one in ManagedClusterSetBinding
> spec:
>   predicates:
>   - requiredClusterSelector:
>       labelSelector:
>         matchExpressions: // From PlacementRule
>          - {key: environment, operator: In, values: ["dev"]}
> ---
> apiVersion: policy.open-cluster-management.io/v1
> kind: PlacementBinding
> metadata:
>   name: binding-policy-example
> placementRef:
>   name: placement-policy-example
>   kind: Placement // Set here to Placement
>   apiGroup: cluster.open-cluster-management.io // Set cluster.open-cluster-management.io
> subjects:
>   - name: policy-example
>     kind: Policy
>     apiGroup: policy.open-cluster-management.io
> ```
> 
> This explanation provides a clear transition from using `PlacementRule` to `Placement`, ensuring compatibility and proper configuration alignment with Open Cluster Management principles.
> 

Reviewer 2: I took what you had written and reflowed the content and pasted it back here (unfortunately the diff doesn't show the changes well 😞 ) since there were some smaller changes that I thought might be easier to show if I pasted the whole thing. The content you have is really good, but there seemed to be some duplication in there that I thought could be streamlined.

````suggestion
## Distributing Policies to Managed Clusters

Distributing a `Policy` to a managed cluster requires four parts, all of which must be in the same namespace:

- [Policy](https://open-cluster-management.io/concepts/policy/): A grouping mechanism for policy templates and is the smallest deployable unit on the hub cluster. Embedded Policy Templates are distributed to applicable managed clusters and acted upon by the appropriate policy controller.
- [PlacementBinding](https://open-cluster-management.io/concepts/policy/#placementbinding): Binds a `Placement` to a `Policy` or `PolicySet`.
- [Placement](https://open-cluster-management.io/concepts/placement/): Dynamically selects a set of `ManagedClusters` in one or multiple `ManagedClusterSet`s.
- [ManagedClusterSetBinding](https://open-cluster-management.io/concepts/managedclusterset/): Binds a `ManagedClusterSet` to a namespace, making this group of managed clusters available for selection by `Placement`.

When the `Policy` is bound to a `Placement` using a `PlacementBinding`, the `Policy` is distributed to the managed clusters and the `Policy` status will report on each cluster returned by the bound `Placement`.

### Using `Placement` with `Policies`

See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for additional details on selecting managed clusters using `Placement`.

To use `Placement` to distribute `Policies`, bind the `Policy` to the `Placement` using a `PlacementBinding`. All of the objects must be in the same namespace. View the following sample `Placement` and `PlacementBinding` bound to a `Policy` named `policy-example` in the namespace `example-ns`. The `Placement` selects managed clusters that have the label `environment: dev`:

```yaml
apiVersion: cluster.open-cluster-management.io/v1beta1
kind: Placement
metadata:
  name: placement-policy-example
  namespace: example-ns
spec:
  predicates:
  - requiredClusterSelector:
      labelSelector:
        matchExpressions:
         - {key: environment, operator: In, values: ["dev"]}
---
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:
  name: binding-policy-example
  namespace: example-ns
placementRef:
  name: placement-policy-example
  kind: Placement
  apiGroup: cluster.open-cluster-management.io
subjects:
  - name: policy-example
    kind: Policy
    apiGroup: policy.open-cluster-management.io
```

### Transitioning from `PlacementRule`(deprecated) to `Placement`

`PlacementRule` had unrestricted access to selecting managed clusters. However, `Placement` requires binding managed clusters to the `Policy` namespace in order for `Policies` to be distributed to those managed clusters, bringing an additional layer of control to system administrators. View the following steps on migrating from `PlacementRule` to `Placement`:

1. The desired managed clusters must be contained in a `ManagedClusterSet`. See the [ManagedClusterSet documentation](https://open-cluster-management.io/concepts/managedclusterset/) to read more, including the default sets that are available that include all managed clusters that would replicate `PlacementRule` behavior.

2. Create a `ManagedClusterSetBinding` to bind the `ManagedClusterSet` to the namespace where you are authoring policies. See the [ManagedClusterSet documentation](https://open-cluster-management.io/concepts/managedclusterset/).

3. Create the `Placement` manifest to replace the `PlacementRule`. To do this, take the selector `spec.clusterSelector` from the `PlacementRule` and put it into `spec.predicates.requiredClusterSelector.labelSelector`. For example, to select managed clusters with the label `environment: dev`:
   ```yaml
   apiVersion: cluster.open-cluster-management.io/v1beta1
   kind: Placement
   metadata:
     name: placement-policy-example
     namespace: example-ns // Ensure this namespace matches the ManagedClusterSetBinding
   spec:
     predicates:
     - requiredClusterSelector:
         labelSelector:
           matchExpressions: // From PlacementRule
           - {key: environment, operator: In, values: ["dev"]}
   ```

   See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for additional details on selecting managed clusters using `Placement`.

4. Identify any `PlacementBinding` resources that reference a `PlacementRule`. Update the `PlacementBinding` to reference the new `Placement`:
   - Change the `placementRef.kind` to `Placement`
   - Update the `placementRef.apiGroup` to `cluster.open-cluster-management.io` to reflect the `Placement`'s API version

   View the following sample `PlacementBinding` that references a `Placement`:

   ```yaml
   apiVersion: policy.open-cluster-management.io/v1
   kind: PlacementBinding
   metadata:
     name: binding-policy-example
   placementRef:
     apiGroup: cluster.open-cluster-management.io // Set to cluster.open-cluster-management.io
     kind: Placement                              // Set to Placement
     name: placement-policy-example
   subjects:
     - name: policy-example
       kind: Policy
       apiGroup: policy.open-cluster-management.io
   ```
````

- [Policy Generator reference YAML](https://github.com/stolostron/policy-generator-plugin/blob/main/docs/policygenerator-reference.yaml)
- [Policy Generator examples](policygenerator)

> ## Transition from PlacementRule to Placement

Reviewer: Can we call this something like: Distributing Policies to Managed Clusters

Then I think there are 3 sections, the Overview of what placement does, the Deprecated PlacementRule details, and then the Placement details.

I think we need to move the text found in another section of the README to the PlacementRules section.  That text is below:

```
The policies are applied to all managed clusters that are available, and have the environment set to dev. If policies need to be applied to another set of clusters, update the PlacementRule.spec.clusterSelector.matchExpressions section in the policies.

Note: As new clusters are added that fit the criteria previously mentioned, the policies are applied automatically.
```

Reviewer 2: I agree, though I think the "overview" can be embedded, like:
```
## Distributing Policies to Managed Clusters

...brief overview/intro about Policy -> PlacementBinding -> Placement and note about PlacementRule being deprecated...

### Using `Placement` with `Policies`

...stuff about Placement...

### Transitioning from `PlacementRule` (deprecated) to `Placement`

...stuff about migrating PlacementRule to Placement...
```

Author: Do we have to write about placement and PlacementRule? because it is already in https://open-cluster-management.io/concepts/placement/.

> ## Transition from PlacementRule to Placement

Reviewer 2: Running Prettier on this file in VSCode ought to clean up/add the expected line breaks like this one:
```suggestion
## Transition from PlacementRule to Placement

```

> ### Prerequisite

Reviewer: I think this section should be called something like: Using PlacementRule to distribute policies (Deprectated)

- [ManagedClusterSet](https://open-cluster-management.io/concepts/managedclusterset/) and [ManagedClusterSetBinding](https://open-cluster-management.io/concepts/managedclusterset)
- `Policy`(The `namespace` in the `policy` should be bound to `ManagedClusterSetBinding`)
- `PlacementBinding` and `PlacementRule`
```yaml
apiVersion: policy.open-cluster-management.io/v1
kind: PlacementBinding
metadata:

>   name: binding-policy-comp-operator

Reviewer 2: (optional) I'd prefer this be more generic.
```suggestion
  name: binding-policy-example
```

placementRef:
  name: placement-policy-comp-operator
  kind: PlacementRule
  apiGroup: apps.open-cluster-management.io
subjects:
  - name: policy-comp-operator
    kind: Policy
    apiGroup: policy.open-cluster-management.io
---
apiVersion: apps.open-cluster-management.io/v1
kind: PlacementRule
metadata:
  name: placement-policy-comp-operator
spec:
  clusterSelector:

>     matchExpressions:
>       - {key: vendor, operator: In, values: ["OpenShift"]}

Reviewer: ```suggestion
    matchExpressions:
      - {key: environment, operator: In, values: ["dev"]}
```

This matches what we documented previously.

```

> ## Convert PlacementRule to Placement

Reviewer: Maybe this should be 2 sections
1. One section that talks about doing the conversion
2. A second section that shows what a user does if they are new and just want to use Placement as a beginner

Maybe so that the conversion section is really brief - it could be the last section.

Author: For second section, should we link to https://open-cluster-management.io/concepts/placement/? Because it is already there. I think rewriting about placement is unnecessary.

Author: I added!

Create the `Placement` manifest as shown below. This `Placement` manifest replaces the `PlacementRule`. Ensure that the `namespace` specified (`comp-operator-ns`) matches the `namespace` used in `ManagedClusterSetBinding`. Copy `matchExpressions` from the `PlacementRule` and paste them into `spec.predicates.requiredClusterSelector.labelSelector`. Update the `PlacementBinding` to reference the new `Placement`. Change the `placementRef.kind` to `Placement` and update the `placementRef.apiGroup` to `cluster.open-cluster-management.io` to reflect the `Placement`'s API version.
```yaml
apiVersion: cluster.open-cluster-management.io/v1beta1
kind: Placement
metadata:
  name: placement-policy-comp-operator
  namespace: comp-operator-ns // Please ensure that this namespace matches one in ManagedClusterSetBinding
spec:
  predicates:
  - requiredClusterSelector:
      labelSelector:
        matchExpressions: // From PlacementRule

>          - {key: vendor, operator: In, values: ["OpenShift"]} 

Reviewer: Change this to match environment=dev like in the other comment.

namespace:

- [Policy](https://open-cluster-management.io/concepts/policy/) is a grouping mechanism for Policy
  Templates and is the smallest deployable unit on the hub cluster. Embedded Policy Templates are
  distributed to applicable managed clusters and acted upon by the appropriate `policy controller`.

- [PlacementBinding](https://open-cluster-management.io/concepts/policy/#placementbinding) binds a
  Placement to a `Policy` or `PolicySet`.

- [Placement](https://open-cluster-management.io/concepts/placement/): Dynamically selects a set of
  `ManagedClusters` in one or multiple `ManagedClusterSet`s.

- [ManagedClusterSetBinding](https://open-cluster-management.io/concepts/managedclusterset/): Binds
  a `ManagedClusterSet` to a namespace, making this group of managed clusters available for
  selection by `Placement`.

When the `Policy` is bound to a `Placement` using a `PlacementBinding`, the `Policy` is distributed
to the managed clusters and the `Policy` status will report on each cluster returned by the bound
`Placement`.

### Using `Placement` with `Policies`

  See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for

>   additional details on selecting managed clusters using `Placement`.

Reviewer 2: This should have stayed over to the left--it's the other paragraph that needed to move over.

   to read more, including the default sets that are available that include all managed clusters
   that would replicate `PlacementRule` behavior.

2. Create a `ManagedClusterSetBinding` to bind the `ManagedClusterSet` to the namespace where you
   are authoring policies. See the
   [ManagedClusterSet documentation](https://open-cluster-management.io/concepts/managedclusterset/).

3. Create the `Placement` manifest to replace the `PlacementRule`. To do this, take the selector
   `spec.clusterSelector` from the `PlacementRule` and put it into
   `spec.predicates.requiredClusterSelector.labelSelector`. For example, to select managed clusters
   with the label `environment: dev`:
   ```yaml
   apiVersion: cluster.open-cluster-management.io/v1beta1
   kind: Placement
   metadata:
     name: placement-policy-example
     namespace: example-ns // Ensure this namespace matches the ManagedClusterSetBinding
   spec:
     predicates:
     - requiredClusterSelector:
         labelSelector:
           matchExpressions: // From PlacementRule
           - {key: environment, operator: In, values: ["dev"]}
   ```

> See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for
> additional details on selecting managed clusters using `Placement`.

Reviewer 2: (Optional) Indent to flow with bullet 3.
```suggestion
   See the [Placement documentation](https://open-cluster-management.io/concepts/placement/) for
   additional details on selecting managed clusters using `Placement`.
```
